File: 0047-permutations-ii/0047-permutations-ii.java
class Solution {
    List<List<Integer>> res;
    List<Integer> path;
    int[] used;
    public List<List<Integer>> permuteUnique(int[] nums) {
        res = new ArrayList<>();
        path = new ArrayList<>();
        used = new int[nums.length];
        Arrays.sort(nums);
        helper(nums);
        return res;
    }

    private void helper(int[] nums) {
        if (path.size() == nums.length) {
            res.add(new ArrayList<Integer>(path));
        }

        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[ i - 1] && used[i - 1] == 1) {
                continue;
            }
            if (used[i] == 0) {
                path.add(nums[i]);
                used[i] = 1;
                helper(nums);
                path.remove(path.size() - 1);
                used[i] = 0;
            }
            
        }
    }
}

File: 0047-permutations-ii/0047-permutations-ii.py
class Solution:
   

    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        path = []
        res = []
        used = [False] * len(nums)
        nums.sort()

        def backtracking():
            if len(path) == len(nums):
                res.append(copy.copy(path))
                return
            
            for i in range(len(nums)):
                if not used[i] and (i == 0 or i > 0 and nums[i] != nums[i - 1] or used[i - 1]):
                    path.append(nums[i])
                    used[i] = True
                    backtracking()
                    used[i] = False
                    path.pop()

        backtracking()
        return res